src/amqp.js

```diff
@@ -0,0 +1,287 @@
+      return await channel.get(queue);
+    } catch (error) {
+      logError(error);
+    }
+  }
+
+  // ACK records
+  async function ackNReplyMessages({status, messages, payloads}) {
+    logger.log('verbose', 'Ack and reply messages!');
+    await messages.forEach((message, index) => {
+      const headers = getHeaderInfo(message);
+
+      // Reply consumer gets: {"data":{"status":"UPDATED","payload":"0"}}
+      sendToQueue({
+        queue: message.properties.correlationId,
+        correlationId: message.properties.correlationId,
+        headers,
+        data: {
+          status, payload: payloads[index]
+        }
+      });
+
+      channel.ack(message);
+    });
+  }
+
+  function ackMessages(messages) {
+    logger.log('verbose', 'Ack messages!');
+    messages.forEach(message => {
+      logger.log('silly', `Ack message ${message.properties.correlationId}`);
+      channel.ack(message);
+    });
+  }
+
+  function nackMessages(messages) {
+    logger.log('verbose', 'Nack messages!');
+    messages.forEach(message => {
+      logger.log('silly', `Nack message ${message.properties.correlationId}`);
+      channel.reject(message, true);
+    });
+  }
+
+  async function sendToQueue({queue, correlationId, headers, data}) {
+    try {
+      logger.log('silly', `Record queue ${queue}`);
+      logger.log('silly', `Record correlationId ${correlationId}`);
+      logger.log('silly', `Record data ${data}`);
+      logger.log('silly', `Record headers ${headers}`);
+
+      await channel.assertQueue(queue, {durable: true});
+
+      channel.sendToQueue(
+        queue,
+        Buffer.from(JSON.stringify({data})),
+        {
+          correlationId,
+          persistent: true,
+          headers
+        }
+      );
+
+      logger.log('silly', `Message send to queue ${queue}`);
+    } catch (error) {
+      logError(error);
+    }
+  }
+
+  async function removeQueue(queue) {
+    await channel.deleteQueue(queue);
+  }
+
+  // ----------------
+  // Helper functions
+  // ----------------
+
+  function messagesToRecords(messages) {
+    logger.log('verbose', 'Parsing messages to records');
+
+    return messages.map(message => {
+      const content = JSON.parse(message.content.toString());
+      return new MarcRecord(content.data);
+    });
+  }
+
+  async function getData(queue) {
+    logger.log('verbose', `Getting queue data from ${queue}`);
+    try {
+      const {messageCount} = await channel.checkQueue(queue);
+      const messagesToGet = messageCount >= CHUNK_SIZE ? CHUNK_SIZE : messageCount;
+
+      const messages = await pump(messagesToGet);
+
+      logger.log('debug', messages);
+
+      return messages;
+    } catch (error) {
+      logError(error);
+    }
+
+    async function pump(count, results = []) {
+      if (count === 0) {
+        return results;
+      }
+
+      const message = await channel.get(queue);
+
+      // Filter not unique messages
+      if (message in results) {
```

Comment:
`in` doesn't work like that: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in

Also, the previous use of `indexOf` wouldn't have worked either since it uses a strict equality check which doesn't work with objects unless you're actually testing for equality of two instances of the same object, but here you want to know if the structure is the same.

Or do you? Why is check done in the first place? Why would RabbitMQ return the same message?

Reply:
Safety against doubles. In older versions there was some cases when rabbit returned same message when async progress was not working. Might be useless now

Reply:
Okay. Now the check is `.includes` which also does strict equality check. Perhaps there is a property for a unique id you can use for checking? Or if not, compare the MARC records `recordA.equalsTo(recordB)`

Reply:
CorrelationId, new implementation uses it.